engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceModificationProcessor.java

```diff
@@ -83,15 +87,45 @@ private void activateElement(
       final AbstractFlowElement elementToActivate) {
     final var elementInstanceRecord = new ProcessInstanceRecord();
     elementInstanceRecord.wrap(processInstance);
+    // todo: deal with non-existing flow scope (#9643)
+    final Optional<Long> flowScopeKey = getFlowScopeKey(processInstance, elementToActivate);
     commandWriter.appendFollowUpCommand(
         keyGenerator.nextKey(),
         ProcessInstanceIntent.ACTIVATE_ELEMENT,
         elementInstanceRecord
-            // todo: allow non-process instance element's as flowscope
-            .setFlowScopeKey(processInstance.getProcessInstanceKey())
+            .setFlowScopeKey(flowScopeKey.get())
             .setBpmnElementType(elementToActivate.getElementType())
             .setElementId(instruction.getElementId())
             .setParentProcessInstanceKey(-1)
             .setParentElementInstanceKey(-1));
   }
+
+  private Optional<Long> getFlowScopeKey(
+      final ProcessInstanceRecord processInstance, final AbstractFlowElement elementToActivate) {
+    final var flowScope = elementToActivate.getFlowScope();
+    if (flowScope.getId().equals(processInstance.getElementIdBuffer())) {
+      return Optional.of(processInstance.getProcessInstanceKey());
+    } else {
+      return getFlowScopeKey(processInstance.getProcessInstanceKey(), flowScope.getId());
+    }
+  }
+
+  private Optional<Long> getFlowScopeKey(
+      final long ancestorKey, final DirectBuffer targetElementId) {
+    final List<ElementInstance> children = elementInstanceState.getChildren(ancestorKey);
+
+    for (final ElementInstance child : children) {
+      if (child.getValue().getElementIdBuffer().equals(targetElementId)) {
+        // todo: instead of early return we should reject the command when multiple matching
+        //  children are found
```

Comment:
We should probably reject only in the `processRecord` method or as close as possible to it, to highlight what the rejection cases are. So we should find a way to move the rejection detections as high as possible up the call tree. 

🔧 If you change the return type of this method to `List` then we can deal with the case of having found multiple higher up the call tree. It also makes it much easier to implement this `getFlowScopeKey` method in function syle, which is likely easier to read.

For example:
```java
final List<Long> matches =
    children.stream()
        .filter(child -> child.getValue().getElementIdBuffer().equals(targetElementId))
        .map(ElementInstance::getKey)
        .toList();

if (!matches.isEmpty()) {
  return matches;
}

return children.stream()
    .map(child -> getFlowScopeKey(child.getKey(), targetElementId))
    .flatMap(Collection::stream)
    .toList();
```

Would allow for the following in the `activate` method:
```java
// todo: deal with non-existing flow scope (#9643)
// todo: deal with multiple flow scopes found without ancestor selection
final List<Long> flowScopeKey = getFlowScopeKey(processInstance, elementToActivate);
commandWriter.appendFollowUpCommand(
..
```

The activate method should probably return an Either when we start dealing with rejections.

Reply:
Also I like your streaming solution way better than mine 😄